### ISSUE TITLE
Refactor Pulse Scheduling Info for more flexibility

### ISSUE THREAD

**Issue by nileshtrivedi** [metabase/metabase#9436]
> under the hood, the way we store pulse scheduling info is fairly tied to our current set of options (daily, weekly, monthly, and fixed set of hour options). In order for us to enable the set of further options we do want to allow, this first needs to change to make sure the scheduling system stays sane.

From: https://github.com/metabase/metabase/issues/3846#issuecomment-463848500

This issue is for the discussion towards this.

**Comment by kdoh** [comment]
Thanks for following up @nileshtrivedi! @camsaul when you get a chance would you mind posting an overview here of what needs to happen to update the backend pulse scheduling code to sanely support scheduling pulses with more timing options like we do for db sync / analyze so that we have an outline for whoever ends up working on this (either external or a team member down the road). 

@nileshtrivedi By opening this are you indicating you'd be interested in working on this refactor yourself?

**Comment by nileshtrivedi** [comment]
@kdoh Yes. Once an approach is decided, @sunitakawane and I will work on it and make it merge-worthy. 👍

**Comment by camsaul** [comment]
Hi @nileshtrivedi, @sunitakawane, as @kdoh mentioned there is a lot that needs to happen to make everything work. Here's a high-level overview of the changes you'll need to make:

1.  Write a DB migration so PulseChannel schedules are stored as cron strings (like we do for DB sync and analyze schedules), instead of the current `schedule_type`/`schedule_hour`/`schedule_day`/`schedule_frame` columns. Migrate everything from the old columns into the new `cron_schedule` column.
2.  Update tests to reflect new columns
2.  Rework the SendPulses job to create one job per Pulse, similar to the SyncDatabases task, scheduled by the cron schedule.
3.  Cancel the old hourly SendPulses job.
4.  Rework the Pulse API endpoints to use the same logic (from `metabase.util.cron`) to translate cron strings into maps readable by the frontend and vice-versa. Update corresponding tests.
5.  Change Pulse UI to use the same scheduling widgets as DB sync and analyze schedules.
6.  Remove old Pulse scheduling widget entirely.

As you can see, it's not really a simple change. I don't really think it's something that I would want a first-time Metabase contributor to try their hand at, so I can't promise I can devote a lot of time to helping you if you get stuck. Good luck

**Comment by sunitakawane** [comment]
Thanks for the details @camsaul . I will start working on this.

**Comment by sunitakawane** [comment]
@camsaul I am working on adding the cron_schedule and need one clarification. Should I add new column `cron_schedule` to `Pulse` or` PulseChannel`. If I add it to `Pulse` , similar to what we have for `Database`, then I will have to add two columns to save `slack` and `email` cron_strings. With `PulseChannel` , I will have to add only one column. Which way is preferred. Please let me know. Thanks

**Comment by sunitakawane** [comment]
@camsaul @mazameli Want to get your feedback on the progress so far,
Here is the PR with tests working [PolyglotNetwork/metabase#3](https://github.com/PolyglotNetwork/metabase/pull/3
)   
So far I have added cron_schedule column to pulse_channels.
I did not remove schedule_type/schedule_hour/schedule_day/schedule_frame columns , since they can be used by frontend to populate the schedules. Also keeping these columns also makes current SendPulses job run as it is. 
Should we add `schedule_min` column to get minute level pulse scheduling? 
Please let us know what you think. Thanks

**Comment by nileshtrivedi** [comment]
Tagging @tlrobinson @kdoh to get this noticed. 😀

**Comment by camsaul** [comment]
@sunitakawane it looks like you missed a bunch of the steps. I don't see:

- [ ] Migrate everything from the old columns into the new `cron_schedule` column.
- [ ] Rework the SendPulses job to create one job per Pulse, similar to the SyncDatabases task, scheduled by the cron schedule.
- [ ] Cancel the old hourly SendPulses job.
- [ ] When a Pulse is deleted its job needs to be canceled. Write tests to confirm this works as expected.
- [ ] When a Pulse's schedule changes its job trigger needs to be updated. Write tests to confirm this works as expected.
- [ ] Rework the Pulse API endpoints to use the same logic (from `metabase.util.cron`) to translate cron strings into maps readable by the frontend and vice-versa. 
- [ ] Change Pulse UI to use the same scheduling widgets as DB sync and analyze schedules.
- [ ] Remove old Pulse scheduling widget entirely.
- [ ] Update any frontend or integration tests related to Pulse scheduling.

Have you tested this? I can't tell by looking at it if you can actually create a new Pulse and schedule it with this PR and have it actually send

**Comment by sunitakawane** [comment]
@camsaul Thanks for reviewing this. As I mentioned in the PR that I did not remove schedule_type/schedule_hour/schedule_day/schedule_frame columns , since they can be used by front end to populate the schedules. And we do not have to migrate everything or rewrite the sendPulse job. 
I did the following steps

- Rework the Pulse API endpoints to use the same logic (from metabase.util.cron) to translate cron strings into maps readable by the frontend and vice-versa.

- Change Pulse UI to use the same scheduling widgets as DB sync and analyze schedules.

I have tested this and it sends pulse as per the schedule. I will work on migration and changing the sendPulses jobs if it is critical. But I am not sure how it will improve the flexibility for scheduling.
It is not clear as to how we will allow more granular pulse scheduling ,since the DB sync widget also has hourly schedule as the most granular scheduling option. Please let me know your thoughts on this.

Thanks
~Sunita Kawane

**Comment by camsaul** [comment]
@sunitakawane the existing SendPulses job runs once an hour and sends any pulses that need to be sent so if you give a Pulse a custom schedule with a cron string for example to send every half hour or  15 minutes after every hour it wouldn't work with the existing job. So right now your PR is a step in the right direction but it won't actually let you schedule Pulses to send whenever you want like people wanted in #3846.

Let me play around with your PR locally and see if I can get some of that stuff quickly.

**Comment by sunitakawane** [comment]
@camsaul you are right about the minute level scheduling and I was thinking of adding `schedule_min` column to get minute level pulse scheduling with the current job. But changing it to what DB_sync is using is a better way to do this. I just wanted to clarify that ,for minute level scheduling , there is more to be done than what you initially listed out. Let me know if you are done reviewing the PR then I can take over the remaining tasks. Thanks

**Comment by camsaul** [comment]
That wouldn't work with the current job unfortunately because the current job runs once an hour regardless.

For minute level scheduling, you would need to do all the things I originally listed, just doing a few of them is not going to work. If you do all of the things then it should work

**Comment by sunitakawane** [comment]
ok, I will work on getting those changes in place and then send the PR. Thanks

**Comment by nileshtrivedi** [comment]
@camsaul Any chance this will be taken up soon? @sunitakawane managed to figure out the change for saving and retrieving cron schedule for pulse. But migrating old pulses and scheduling pulse job changes are turning out to be a bit tricky for us beginners.

**Comment by dhk** [comment]
Hey team: we @ Syncera would love to get a more granular schedule than hourly. We have important jobs that are time sensitive and to wait an hour is less than desirable.

What's the status on this?

**Comment by alexyarosh** [comment]
Starting v55, [question alerts ](https://www.metabase.com/docs/latest/questions/alerts)(but not yet dashboard subscriptions) can be configured with custom cron schedule 🎉
